// @temporalio/workflow/commonjs/time.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nullToUndefined = exports.msOptionalStrToTs = exports.msStrToTs = exports.msToTs = exports.tsToMs = void 0;
const long_1 = __importDefault(require("long"));
const ms_1 = __importDefault(require("ms"));
/**
 * Lossy conversion function from Timestamp to number due to possible overflow
 */
function tsToMs(ts) {
    if (ts === undefined || ts === null) {
        throw new Error(`Expected timestamp, got ${ts}`);
    }
    const { seconds, nanos } = ts;
    return (seconds || long_1.default.UZERO)
        .mul(1000)
        .add(Math.floor((nanos || 0) / 1000000))
        .toNumber();
}
exports.tsToMs = tsToMs;
function msToTs(millis) {
    const seconds = Math.floor(millis / 1000);
    const nanos = (millis % 1000) * 1000000;
    return { seconds: long_1.default.fromNumber(seconds), nanos };
}
exports.msToTs = msToTs;
function msStrToTs(str) {
    return msToTs(ms_1.default(str));
}
exports.msStrToTs = msStrToTs;
function msOptionalStrToTs(str) {
    if (str === undefined)
        return undefined;
    return msToTs(ms_1.default(str));
}
exports.msOptionalStrToTs = msOptionalStrToTs;
function nullToUndefined(x) {
    if (x === null)
        return undefined;
    return x;
}
exports.nullToUndefined = nullToUndefined;
