// @temporalio/workflow/commonjs/converter/payload-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BinaryPayloadConverter = exports.JsonPayloadConverter = exports.UndefinedPayloadConverter = void 0;
const types_1 = require("./types");
/**
 * Converts between JS undefined and NULL Payload
 */
class UndefinedPayloadConverter {
    constructor() {
        this.encodingType = types_1.encodingTypes.METADATA_ENCODING_NULL;
    }
    toData(value) {
        if (value !== undefined)
            return undefined; // Can't encode
        return {
            metadata: {
                [types_1.METADATA_ENCODING_KEY]: types_1.encodingKeys.METADATA_ENCODING_NULL,
            },
        };
    }
    fromData(_content) {
        return undefined; // Just return undeinfed
    }
}
exports.UndefinedPayloadConverter = UndefinedPayloadConverter;
/**
 * Converts between non-undefined values and serialized JSON Payload
 */
class JsonPayloadConverter {
    constructor() {
        this.encodingType = types_1.encodingTypes.METADATA_ENCODING_JSON;
    }
    toData(value) {
        if (value === undefined)
            return undefined; // Should be encoded with the UndefinedPayloadConverter
        return {
            metadata: {
                [types_1.METADATA_ENCODING_KEY]: types_1.encodingKeys.METADATA_ENCODING_JSON,
            },
            data: types_1.u8(JSON.stringify(value)),
        };
    }
    fromData(content) {
        if (content.data === undefined || content.data === null) {
            throw new types_1.ValueError('Got payload with no data');
        }
        return JSON.parse(types_1.str(content.data));
    }
}
exports.JsonPayloadConverter = JsonPayloadConverter;
/**
 * Converts between binary data types and RAW Payload
 */
class BinaryPayloadConverter {
    constructor() {
        this.encodingType = types_1.encodingTypes.METADATA_ENCODING_RAW;
    }
    toData(value) {
        // TODO: support any DataView or ArrayBuffer?
        if (!(value instanceof Uint8Array)) {
            return undefined;
        }
        return {
            metadata: {
                [types_1.METADATA_ENCODING_KEY]: types_1.encodingKeys.METADATA_ENCODING_RAW,
            },
            data: value,
        };
    }
    fromData(content) {
        // TODO: support any DataView or ArrayBuffer?
        return content.data;
    }
}
exports.BinaryPayloadConverter = BinaryPayloadConverter;
