// @temporalio/worker/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GracefulShutdownPeriodExpiredError = exports.UnexpectedError = exports.WorkflowError = exports.TransportError = exports.ShutdownError = void 0;
/**
 * An unhandled error while communicating with the server, considered fatal
 */
class ShutdownError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'ShutdownError';
    }
}
exports.ShutdownError = ShutdownError;
/**
 * Thrown after shutdown was requested as a response to a poll function, JS should stop polling
 * once this error is encountered
 */
class TransportError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'TransportError';
    }
}
exports.TransportError = TransportError;
/**
 * Workflow did something Core did not expect, it should be immediately deleted from the cache
 */
class WorkflowError extends Error {
    constructor(message, runId, source) {
        super(message);
        this.runId = runId;
        this.source = source;
        this.name = 'WorkflowError';
    }
}
exports.WorkflowError = WorkflowError;
/**
 * Something unexpected happened, considered fatal
 */
class UnexpectedError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'UnexpectedError';
    }
}
exports.UnexpectedError = UnexpectedError;
/**
 * Thrown from JS if Worker does not shutdown in configured period
 */
class GracefulShutdownPeriodExpiredError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'GracefulShutdownPeriodExpiredError';
    }
}
exports.GracefulShutdownPeriodExpiredError = GracefulShutdownPeriodExpiredError;
