// @temporalio/workflow/es2020/errors.js
export class WorkflowExecutionTerminatedError extends Error {
    constructor(message, details, identity) {
        super(message);
        this.details = details;
        this.identity = identity;
        this.name = 'WorkflowExecutionTerminatedError';
    }
}
export class WorkflowExecutionTimedOutError extends Error {
    constructor(message, retryState) {
        super(message);
        this.retryState = retryState;
        this.name = 'WorkflowExecutionTimedOutError';
    }
}
export class WorkflowExecutionFailedError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'WorkflowExecutionFailedError';
    }
}
export class WorkflowExecutionCancelledError extends Error {
    constructor(message, details) {
        super(message);
        this.details = details;
        this.name = 'WorkflowExecutionCancelledError';
    }
}
export class WorkflowExecutionContinuedAsNewError extends Error {
    constructor(message, newExecutionRunId) {
        super(message);
        this.newExecutionRunId = newExecutionRunId;
        this.name = 'WorkflowExecutionContinuedAsNewError';
    }
}
/**
 * Thrown in workflow when it is requested to be cancelled either externally or internally.
 *
 * @see {@link CancellationSource}
 */
export class CancellationError extends Error {
    constructor(message, source) {
        super(message);
        this.source = source;
        this.name = 'CancellationError';
    }
}
/**
 * Used in different parts of the project to signal that something unexpected has happened
 */
export class IllegalStateError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'IllegalStateError';
    }
}
/**
 * Thrown in workflow when it trys to do something that non-deterministic such as construct a WeakMap()
 */
export class DeterminismViolationError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'DeterminismViolationError';
    }
}
